src/flow.ts

```diff
@@ -606,6 +614,65 @@ export class Flow {
     localFlags[index] = flags & ~flag;
   }
 
+  /** set type assign */
+  setAssignType(expr: ExpressionRef, element: TypedElement, type: Type): void {
+    if (type && !type.isReference) return;
+    this.setNarrowedType(element, type);
+    this.conditionalNarrowedType.setAssignType(expr, element, type);
+  }
+  /** set type narrow */
+  setNarrowedType(element: TypedElement, type: Type): void {
+    if (!type.isReference) return;
+    if (this.narrowedTypes == null) {
+      this.narrowedTypes = new NarrowedTypeMap();
+    }
+    let narrowedTypes = assert(this.narrowedTypes);
```

Comment:
Also `assert` should be unnecessary

Reply:
If type narrow does not effect for Field, It should like this:
```typescript
let narrowedTypes = this.narrowedTypes || new NarrowedTypeMap();
this.narrowedTypes = narrowedTypes;
```